### DISCUSSION TITLE
함수형 인터페이스(Functional Interface)에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by kunsanglee** [2024-woowacourse-study/level-interview#64] [Java]
.

**Comment by jimi567** [comment]
함수형 인터페이스는 딱 하나의 추상 메서드를 가지고 있는 인터페이스를 뜻한다. 기본 함수형 인터페이스로는  Runnable, Comparator, Piredicate 등이 있다.
람다 표현식이나 메서드 참조와 같은 함수형 프로그래밍의 스타일을 지원하기 위해 사용된다.

병렬처리 및 비동기 프로그래밍을 할 때, 유용하며, 코드의 재사용성을 높혀주며,  무엇보다 코드의 가독성을 높이고 간결하게 작성할 수 있는 장점이 있다.

또한, @FunctionalInterface 어노테이션을 통해 명시적으로 함수형 인터페이스를 선언할 수 있으며, 이를 통해 컴파일러가 해당 인터페이스가 함수형 인터페이스임을 검증해주기도 한다.

**Comment by nayonsoso** [comment]
함수형 인터페이스란, `하나의 추상 메서드만을 갖는 인터페이스`입니다.
함수형 인터페이스를 사용하면 람다 표현식을 쓸 수 있는데요,
이는 함수형 인터페이스 내부에 단 하나의 추상 메서드만 존재하기 때문에 
컴파일러가 메서드의 시그니처를 추론할 수 있기 때문입니다.
람다를 사용하면 코드를 간결하게 쓸 수 있다는 장점이 있습니다.

`@FunctionalInterface` 는 인터페이스가 함수형 인터페이스임을 명시하기 위해 사용합니다.
`@FunctionalInterface`를 사용하면 인터페이스에 불필요한 메서드가 존재할 때 컴파일 에러가 발생합니다.
